Instructions to run a multi-node subnet (#146)
consensus-shipyard/ipc-agent: README.md

# Copy the wallet key inside the container
$ docker exec -it <container-id> sh -c "./eudico wallet import --format=json-lotus <target-file-in-container>"
# Sample execution
$ docker cp ~/.ipc-agent/t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy.key 91d2af805346:/input.key
$ docker exec -it 91d2af805346 sh -c "eudico wallet import --format=json-lotus input.key"
```

### Running a subnet with several validators
In this section, we will deploy a subnet where the IPC agent is responsible for handling more than one validator in the subnet. Throughout this guide, we use the `ipc-infra` scripts to deploy the nodes in Docker containers, but a similar steps could be used to deploy the nodes on bare-metal or any other runtime.

For the rest of this tutorial, we'll assume that you have your agent already configured and interacting with a rootnet. For more information on how to connect to a rootnet [revisit this section](#Interacting-with-a-rootnet). We are going to deploy a subnet with 5 validators. The first thing we'll need to do is creating a new wallet for every validator we want to run. We can do this directly through the agent with the following command:
```bash

> $ ./bin/ipc-agent wallet-new --key-type=secp256k1 --subnet=/root

Author: Update: I just synced with @cryptoAtwill and we will try to include a nice refactor in the next release :pray:

```
We also need to provide with some funds our wallets so they can put collateral to join the subnet. According to the rootnet you are connected to, you may need to get some funds from the faucet, or send some from your main wallet. Funds can be send from your main wallet also through the agent with: 
```bash
$ ./bin/ipc-agent send-value --subnet=/root --to=<target-wallet> <amount_FIL>
```
With this, we can already create the subnet with `/root` as its parent. We are going to set the `--min-validators 5` so no new blocks can be created without this number of validators in the subnet.
```bash
# Creating a sample subnet with 5 as the minimum number of validators.

> ./bin/ipc-agent create-subnet -p /root -n test --min-validator-stake 1 --min-validators 5 --finality-threshold 10 --check-period 10

Author: Only registered subnets with enough collateral appear in `list-subnets`. I will update the command description and add a note to the README

```

#### Deploying the infrastructure
In order to deploy the 5 validators for the subnet, we will have to first export the keys from our root node so we can import them to our validators. Depending on how you are running your rootnet node you'll have to make a call to the docker container, or your nodes API. More information about exporting keys from your node can be found under [this section](#Exporting-wallet-keys).

With the five keys conveniently exported, we can deploy the subnet nodes using the `infra-scripts`. The following code snippet showcases the deployment of five sample nodes. Note that each node should be importing a different wallet key for their validator, and should be exposing different ports for their API and validators:
```bash

> $ ./bin/ipc-infra/run-subnet-docker.sh 1240 1359 /root/t01002 ~/.ipc-agent/wallet1.key

Reviewer: I think we can move the guide part into docs.filecoin.io ([right now it's a clone](https://interplanetary-consensus.netlify.app/basics/interplanetary-consensus/overview/)) while keeping this as the readme. It makes sense in my mind, at least.

$ ./bin/ipc-infra/run-subnet-docker.sh 1250 1369 /root/t01002 ~/.ipc-agent/wallet2.key
$ ./bin/ipc-infra/run-subnet-docker.sh 1280 1379 /root/t01002 ~/.ipc-agent/wallet3.key
$ ./bin/ipc-infra/run-subnet-docker.sh 1270 1389 /root/t01002 ~/.ipc-agent/wallet4.key

> $ ./bin/ipc-infra/run-subnet-docker.sh 1280 1399 /root/t01002 ~/.ipc-agent/wallet5.key

Author: Thank goodness that I have you as my proof-reader :facepalm:

```
>>> Subnet /root/t01002 daemon running in container: 91d2af80534665a8d9a20127e480c16136d352a79563e74ee3c5497d50b9eda8 (friendly name: ipc_root_t01002_1240)
>>> Token to /root/t01002 daemon: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.JTiumQwFIutkTb0gUC5JWTATs-lUvDaopEDE0ewgzLk
>>> Default wallet: t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy
>>> Subnet subnet validator info:
t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy@/ip4/172.17.0.4/udp/1348/quic/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy@/ip4/127.0.0.1/udp/1348/quic/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy@/ip4/172.17.0.4/tcp/1347/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy@/ip4/127.0.0.1/tcp/1347/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
>>> API listening in host port 1240
>>> Validator listening in host port 1359
```

#### Configuring the agent

> To configure the agent for its use with all the validators, we need to connect to the RPC API of one of the validators, and import all of the wallets of the validators in that node, so the agent is able through the same API to act on behalf of any validator. More information about importing [keys can be found in this section](#Importing-a-wallet-to-a-subnet-node).

Reviewer: > Ideally, the agent would have the ability to host the wallets, sign messages locally, and then forward the message signed to Lotus for submission through the RPC API of any of the validators, but we haven't prioritized this,

I think we're good for now, but this definitely seems like an M2 item

Here's an example of the configuration connecting to the RPC of the first validator, and configuring all the wallets for the validators in the subnet.
```toml
[[subnets]]
id = "/root/t01002"
gateway_addr = "t064"
network_name = "test"
jsonrpc_api_http = "http://127.0.0.1:1240/rpc/v1"
auth_token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.JTiumQwFIutkTb0gUC5JWTATs-lUvDaopEDE0ewgzLk"
accounts = ["t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy", "t1cp4q4lqsdhob23ysywffg2tvbmar5cshia4rweq", "t1nv5jrdxk4ljzndaecfjgmu35k6iz54pkufktvua", "t1pr3qzqieikp33pfdaygwkwt5v35a5ljsxsu64xq", "t1pmxk3nhg2u2ge7ioilrk7c5rmdmmbgpazipvzyy"]
```
Remember to run `./bin/ipc-agent reload-config` for your agent to pick up the latest changes for the config.

#### Joining the subnet

> All the infrastructure for the subnet is now deployed, and we can join our validators to the subnet. For this, we need to send a `join` command from each of our validators from their validator wallet addresses providing the validators multiaddress. We can get the validator multiaddress from the output of the script we ran to deploy the infrastructure (or by running `eudico mir validator config validator-addr`). These scripts expose through docker the port for the TPC port of the libp2p host of the validator (if there is interest to use the UDP-based multiaddress drop an issue and we can update them accordingly). Hence, we need to choose from the available multiaddresses one that exposes a TCP connection. Finally, our validators are configured so their docker containers interact with each other through ports in the host machine, so instead of exposing an IP address for the multiaddress, the following DNS-based multiaddress needs to be used for all validators: `/dns/host.docker.internal/`.

Reviewer: I'll take your word for it but, in the previous tests, things always worked for me with IP address. There's probably a way to make either work, but, as it stands, I don't think the instructions do, mainly because eg

```
>>> Subnet /root/t01015 daemon running in container: 975db017f41c41ca423fa04205add5f89bc28a8c68c4eadac06e044d0e8d3c5f (friendly name: ipc_root_t01015_1240)
>>> Token to /root/t01015 daemon: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0.GjgfYovLNSkVZmO7AoVu-YJeqv1A_s0BvZSsR54xtEg
>>> Default wallet: t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q
>>> Subnet subnet validator info:
t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/ip4/172.17.0.2/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE
t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/ip4/127.0.0.1/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE
t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/ip4/172.17.0.2/udp/1348/quic/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE
t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/ip4/127.0.0.1/udp/1348/quic/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE
>>> API listening in host port 1240
>>> Validator listening in host port 1359
```

The multiaddr uses the same ports for every node (1347, 1348) and not the docker exposed ports (1240, 1359). So if you take the addresses output by eudico and just replace the IP address with `/dns/host.docker.internal/`, you'll get, for every node:

` --validator-net-addr="/dns/host.docker.internal/tcp/1347/p2p/<peer id>"`

Which evidently can't work. So we need to either tell people to use the right ports or the right address.

Reviewer: Related potential bad idea... Should we change the deployment script again to filter the validator-addr output and replace it to yield the output people will actually need? Something like...

```
docker exec -it $img eudico mir validator config validator-addr | grep '127.0.0.1/tcp' | sed 's/^.*\/127.0.0.1/\/dns\/host.docker.internal/' | sed -E "s/\/[0-9]+\//\/$PORT\//"
```

yielding example output

```
/dns/host.docker.internal/tcp/1240/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE
```

instead of the current 4 lines

🤷

Author: > 'll take your word for it but, in the previous tests, things always worked for me with IP address.
This is because you were running a subnet with a single validator that doesn't need to dial itself, so as long as you provide a valid multiaddr for its peerID it will work, but if you run all validators on a docker container and try to dial other containers it'll fail. 

I agree with all of your points, and this one is also a hard one. I wanted to strike the right balance between completely abstracting everything from the user and giving enough information for someone not using Docker and trying to make a production deployment to know what is happening under the hood. You renaming proposals is probably the best approach (and the one I like the most). That way users can just copy-paste the address and that's it. Do you mind adding an issue so we tackle in the next iteration? (that way I can merge this one and cut a release before we start the next milestone). Thanks :pray:

Reviewer: I can, but I'd still really like to run the multinode deployment to successful completion. I'm in back-to-back meetings but will try to do it during the day. Or you can merge and I'll provide feedback afterwards. What's your ETA for the release?

Author: I can wait, no rush. Feel free to ping me if you want to sync and unblock this.

Reviewer: PR here: https://github.com/consensus-shipyard/lotus/pull/145

Not strongly attached to it, feel free to merge or close at your convenience.

This is the command that needs to be executed for every validator to join the subnet:
```bash
$ ./bin/ipc-agent join-subnet --from=<validator-wallet> --subnet=/root/t01002 --collateral=<amount-collateral> --validator-net-addr="/dns/host.docker.internal/tcp/<VALIDATOR_PORT>/p2p/<VALIDATOR_MULTIADDR>"

# Sample execution for the validator whose logs where shared above for 2FIL collateral
$ ./bin/ipc-agent join-subnet --from=t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy --subnet=/root/t01002 --collateral=2 --validator-net-addr="/dns/host.docker.internal/tcp/1359/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR
```
Remember doing the above step for the five validators.

#### Mining in subnet
We have everything in place now to start mining from all the validators. Mining is as simple as running the following script passing the container id for the validators:
```bash

> $  ./bin/ipc-infra/mine-subnet.sh <node-container-id>

Reviewer: Failed, just blocks here:

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-infra/mine-subnet.sh 975db017f41c41ca423fa04205add5f89bc28a8c68c4eadac06e044d0e8d3c5f
[*] Starting to mine on validator
2023-04-01T16:13:05.139Z        INFO    mir-validator-cli       mirvalidator/run.go:143 Checking full node sync status
2023/04/01 16:13:05 failed to sufficiently increase receive buffer size (was: 208 kiB, wanted: 2048 kiB, got: 416 kiB). See https://github.com/lucas-clemente/quic-go/wiki/UDP-Receive-Buffer-Size for details.
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:162 Mir libp2p host listening in the following addresses:
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/172.17.0.2/udp/1348/quic
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/127.0.0.1/udp/1348/quic
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/172.17.0.2/tcp/1347
2023-04-01T16:13:05.144Z        INFO    mir-validator-cli       mirvalidator/run.go:164 /ip4/127.0.0.1/tcp/1347
2023-04-01T16:13:05.157Z        INFO    mir-validator-cli       mirvalidator/run.go:228 Starting mining with validator  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:183 stored next configuration number not found      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:195 stored executed configuration number not found  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.169Z        INFO    mir-consensus   mir/conf_manager.go:208 stored reconfiguration votes not found  {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.170Z        INFO    mir-consensus   mir/manager.go:244      Mir manager serve started       {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.170Z        INFO    mir-consensus   mir/manager.go:248      Mir info:
        Network - /root/t01015
        Validator ID - t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q
        Mir peerID - t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q
        Validators - [t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1 t14obklze7ke4ccepzceo5reqb2kexe4xuskprhlq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWAujMYtMdDXQyq7dPMRHT6pzJo7SDpNEyQwDi5YM21xLE t1qaujcu3rs56sxt4f3ikapzknenvajdvfgc7g3gq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWEDveetd6KRgrdjSjTEKXfe54zKNdRv3DAvVJaChMQr3P t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf]   {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:220        RestoreState for epoch 0 started        {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:244        RestoreState: next membership size is 5 at epoch 0      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:276        Snapshot len is zero    {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:279        RestoreState for epoch 0 finished       {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:463        New epoch started: updating 0 to 0      {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
2023-04-01T16:13:05.171Z        INFO    mir-consensus   mir/state_manager.go:489        New epoch finished: updating 0 to 0     {"validator": "t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q"}
```

Maybe on account of validator addresses -- see above?

Reviewer: The bigger issue is that the user gets stuck here too... I assume I have the wrong validator address. How do I fix it?

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent leave-subnet --subnet /root/t01015
[2023-04-01T16:20:18Z ERROR ipc_agent::cli::commands] process command: LeaveSubnet(LeaveSubnetArgs { ipc_agent_url: None, from: None, subnet: "/root/t01015" }) failed due to error: json_rpc error: {"code":-1,"data":null,"message":"json_rpc error: {\"code\":1,\"message\":\"GasEstimateMessageGas error: message execution failed: exit 20, reason: message failed with backtrace:\\n00: f01015 (method 79496443) -- caller has no stake in subnet (20)\\n (RetCode=20)\"}"}
```

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent set-validator-net-addr
error: the following required arguments were not provided:
  --subnet <SUBNET>
  --validator-net-addr <VALIDATOR_NET_ADDR>

Usage: ipc-agent set-validator-net-addr --subnet <SUBNET> --validator-net-addr <VALIDATOR_NET_ADDR>
```

Now, this will naturally error out without it as my default wallet isn't a validator and that ` from: None` in the error suggests I should be able to pass a `--from=` to say which validator I want to leave (and, indeed, I can, both for leave-subnet and set-validator-net-addr). But the message & usage instructions don't really mention that possibility and I also didn't notice it at first. So, instead, I just tried to join again with each validator, and it seemed to work:

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc-agent join-subnet --from=t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi --subnet=/root/t01015 --collateral=1 --validator-net-addr="/ip4/172.17.0.6/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf"
[2023-04-01T16:25:58Z INFO  ipc_agent::cli::commands::manager::join] joined subnet: /root/t01015
```

Except that it didn't. It still gets stuck and the validator list by mine-subnet still shows:
```
        Validators - [t1aa4er7k6c7jjcv6vmdz7wtm76qhj3x3qp4ltj2q@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWMwhd7SREGYj3ZSSWRdtdC4Y51dxdWBF3UZgQXzWb9PQE t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1 t14obklze7ke4ccepzceo5reqb2kexe4xuskprhlq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWAujMYtMdDXQyq7dPMRHT6pzJo7SDpNEyQwDi5YM21xLE t1qaujcu3rs56sxt4f3ikapzknenvajdvfgc7g3gq@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWEDveetd6KRgrdjSjTEKXfe54zKNdRv3DAvVJaChMQr3P t1erhas3yxe7njdlhv3zcm4mzk2wlh56vr5l3czdi@/dns/host.docker.internal/tcp/1347/p2p/12D3KooWLVkXDSk6WteGkUsY2Roi7w8yrESLGJm35SuuUZF51Wzf]
```

So then I passed from to `set-validator-net-addr` for each node, and it went through, i.e. now shows the right set. Still doesn't run, though.

So I tried to restart the containers and `set-validator-net-addr` again, but now I just get this

```
ubuntu@ip-172-26-28-130:~/ipc/ipc-agent$ ./bin/ipc_agent set-validator-net-addr --subnet=/root/t01015 --from=t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty --validator-net-addr=/dns/host.docker.internal/tcp/1250/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1
[2023-04-01T16:56:17Z ERROR ipc_agent::cli::commands] process command: SetValidatorNetAddr(SetValidatorNetAddrArgs { ipc_agent_url: None, from: Some("t1pufa2fr5yh6rpasqiedpvbwnpx6lgwtx5h7n5ty"), subnet: "/root/t01015", validator_net_addr: "/dns/host.docker.internal/tcp/1250/p2p/12D3KooWFa918fedPkDpTibdrisiPGFpWU1Fjbs2JzuXMWwtr7s1" }) failed due to error: No such file or directory (os error 2)
```

Other commands work and the error is very unhelpful. Inspecting the running docker containers also shows errors but not a lot of directions forward, so I guess I'll just have to start from scratch. But no time to do it now :(

Author: > Failed, just blocks here:
This is a validator waiting for the rest of validators to come online. Sometimes the first block requires several rounds of reconfiguration to be produced so it may take ~1min with the current block times. If there was an error on the configuration of the validator's addresses you should have started seeing dialing errors. 

That being said, this interaction is SUPER useful, thank you for guiding me through your reasoning. It is a perfect example of how horrible the UX is and how much it needs to be improved. I have some ideas of things we could improve. I'll start tracking issues :pray:

Reviewer: So the issue was that I wasn't starting mining in each of the validators, self-misled by a previous discussion on control. Instructions updated slightly.

This is the command that needs to be executed for every validator to join the subnet:
```bash
$ ./bin/ipc-agent join-subnet --from=<validator-wallet> --subnet=/root/t01002 --collateral=<amount-collateral> --validator-net-addr="/dns/host.docker.internal/tcp/<VALIDATOR_PORT>/p2p/<VALIDATOR_MULTIADDR>"

# Sample execution for the validator whose logs where shared above for 2FIL collateral
$ ./bin/ipc-agent join-subnet --from=t1ivy6mo2ofxw4fdmft22nel66w63fb7cuyslm4cy --subnet=/root/t01002 --collateral=2 --validator-net-addr="/dns/host.docker.internal/tcp/1359/p2p/12D3KooWEJXcSPw6Yv4jDk52xvp2rdeG3J6jCPX9AgBJE2mRCVoR"
```
Remember doing the above step for the five validators.

#### Mining in subnet

> We have everything in place now to start mining from all the validators. Mining is as simple as running the following script passing the container id for the validators:

Reviewer: ```suggestion
We have everything in place now to start mining. Mining is as simple as running the following script for each of the validators, passing the container id/name:
```
